src/com/amazon/ion/impl/bin/WriteBuffer.java

```diff
@@ -1266,10 +1255,14 @@ public void writeUInt8At(final long position, final long value)
     {
         final int index = index(position);
         final int offset = offset(position);
-
         // XXX we'll never overrun a block unless we're given a position past our block array
         final Block block = blocks.get(index);
-        block.data[offset] = (byte) value;
+        long bitValue = block.data[offset];
+        if (value <= 0xD) {
```

Comment:
I suggest creating a new method (something like 'writeLowerNibbleAt'), and call it for this purpose, instead of branching within this method.